### PR TITLE
Add CLIP score

### LINKED ISSUE #1311: Add CLIP score
## 🚀 Feature

Calculate the correlation between image and text
### Motivation

Evaluating the performance of the text2image model
### Pitch

pytorch-like Pseudocode
```python
def clip_score(img_inputs, txt_inputs):
    img_features = clip.get_image_features(img_inputs)
    txt_features = clip.get_text_features(txt_inputs)
    img_features, txt_features = [
        x / torch.linalg.norm(x, axis=-1, keepdims=True)
        for x in [img_features, txt_features]
    ]
    return (img_features * txt_features).sum(axis=-1)
```
### Additional context

[clip score](https://wandb.ai/dalle-mini/dalle-mini/reports/CLIP-score-vs-FID-pareto-curves--VmlldzoyMDYyNTAy)

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1311
Adds [CLIP score](https://arxiv.org/pdf/2104.08718.pdf) which will be our first multi modal metric :D

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/index.rst [hunk 1]
```diff
@@ -158,6 +158,14 @@ Or directly from conda
 
    image/*
 
+.. toctree::
```
Comment (stancld) [inline docs/source/index.rst:161]: We've sorted this section in an alphabetical order in #1298, so maybe we can keep it here?

File: docs/source/multimodal/clip_score.rst [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+   :tags: Image
+
+############################################
+CLIP Score
+############################################
+
+Module Interface
+________________
+
+.. autoclass:: torchmetrics.multimodal.clip_score.CLIPScore
```
Comment (stancld) [inline docs/source/multimodal/clip_score.rst:13]: Shouldn't we also have `functional` metric?

File: src/torchmetrics/multimodal/clip_score.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    from transformers import CLIPModel as _CLIPModel
+    from transformers import CLIPProcessor as _CLIPProcessor
+else:
+    __doctest_skip__ = ["CLIPScore"]
+
+from torchmetrics import Metric
+
+
+class CLIPScore(Metric):
+    """`CLIP Score`_ is a reference free metric that can be used to evaluate the correlation between an generated
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_score.py:32]:
```suggestion
    """`CLIP Score`_ is a reference free metric that can be used to evaluate the correlation between a generated
```

File: src/torchmetrics/multimodal/clip_score.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        \text{CLIPScore(I, C)} = max(100 * cos(E_I, E_C), 0)
+
+    which corresponds to the cosine similarity between visual CLIP embedding :math:`E_i` for an image :math:`i` and
+    textual CLIP embedding :math:`E_C` for an caption :math:`C`. The score is bound between 0 and 100 and the closer
+    to 100 the better.
+
+    .. note:: Metric is not scriptable
+
+    Args:
+        version: string indicating the version of the CLIP model to use. Available models are
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_score.py:46]: We use `model_name_or_path` for `BertScore`. Maybe can we diverge here from the reference package and keep our naming which is consistent with `transformers` package?

File: src/torchmetrics/multimodal/clip_score.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                " Either install with `pip install transformers>=4.0` or `pip install torchmetrics[multimodal]`."
+            )
+        self.add_state("score", torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("n_samples", torch.tensor(0, dtype=torch.long), dist_reduce_fx="sum")
+
+    def update(self, images: Union[Tensor, List[Tensor]], text: Union[str, List[str]]) -> None:
+        """Updates CLIP score on a batch of images and text.
+
+        Args:
+            images: Either a single [N, C, H, W] tensor or an list of [C, H, W] tensors
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_score.py:97]:
```suggestion
            images: Either a single [N, C, H, W] tensor or a list of [C, H, W] tensors
```

File: src/torchmetrics/multimodal/clip_score.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            text: Either a single caption or a list of captions
+
+        Raises:
+            ValueError:
+                If not all images have format [C, H, W]
+            ValueError:
+                If the number of images and captions do not match
+        """
+        if not isinstance(images, List):
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_score.py:106]:
```suggestion
        if not isinstance(images, list):
```

File: src/torchmetrics/multimodal/clip_score.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            if images.ndim == 3:
+                images = [images]
+            else:  # unwrap into list
+                images = [i for i in images]
+
+        if not all(i.ndim == 3 for i in images):
+            raise ValueError("Expected all images to be 3d but found image that has either more or less")
+
+        if not isinstance(text, List):
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_score.py:115]:
```suggestion
        if not isinstance(text, list):
```

File: tests/unittests/multimodal/test_clip_score.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        _random_input,
+    ],
+)
+class TestCLIPScore(MetricTester):
+    atol = 1e-5
+
+    @pytest.mark.parametrize("ddp", [True, False])
+    def test_clip_score(self, input, version, ddp):
+        # images are preds and targets are captions
+        preds, target = input
```
Comment (stancld) [inline tests/unittests/multimodal/test_clip_score.py:66]:
```suggestion
    "input",
    [
        _random_input.preds, _random_input.target
    ],
)
class TestCLIPScore(MetricTester):
    atol = 1e-5

    @pytest.mark.parametrize("ddp", [True, False])
    def test_clip_score(self, preds, target, version, ddp):
        # images are preds and targets are captions
```

To stay consistent with other tests?

+ Shouldn't we add `dist_sync_on_step = True | False` into the configuration?

File: tests/unittests/multimodal/test_clip_score.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            ddp=ddp,
+            preds=preds,
+            target=target,
+            metric_class=CLIPScore,
+            sk_metric=partial(_compare_fn, version=version),
+            metric_args={"version": version},
+            check_scriptable=False,
+            check_state_dict=False,
+        )
+
```
Comment (stancld) [inline tests/unittests/multimodal/test_clip_score.py:77]: Shouldn't we add a `functional` metric and then also `differentiability` test?

**Review by stancld — COMMENTED** [review]
Cool to have the first multimodal metric in `torchmetrics`. Just a few nitpicks, otherwise great work! :+1:

File: tests/unittests/multimodal/test_clip_score.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+@pytest.mark.parametrize("version", ["openai/clip-vit-base-patch32"])
+@pytest.mark.parametrize(
+    "input",
+    [
+        _random_input,
+    ],
+)
+class TestCLIPScore(MetricTester):
```
Comment (stancld) [inline tests/unittests/multimodal/test_clip_score.py:60]: We should mark the tests to be skip if we fail to download `model` from HF hub, similarly as we do for `BERTScore`

**Comment by SkafteNicki** [comment]
@Borda what to do here?
The tests are running out of memory because the underlying CLIP model that is used to calculate the metric is around ~600 mb in size.

**Comment by Borda** [comment]
> @Borda what to do here? The tests are running out of memory because the underlying CLIP model that is used to calculate the metric is around ~600 mb in size.

how about running them only on GPU? :rabbit:

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/functional/multimodal/clip_score.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+def _get_model_and_processor(
+    model_name_or_path: Literal[
+        "openai/clip-vit-base-patch16",
+        "openai/clip-vit-base-patch32",
+        "openai/clip-vit-large-patch14-336",
+        "openai/clip-vit-large-patch14",
+    ] = "openai/clip-vit-large-patch14",
+) -> Tuple[_CLIPModel, _CLIPProcessor]:
+    if _TRANSFORMERS_AVAILABLE:
```
[the remark this answers is not in the document]
Comment (stancld) [reply src/torchmetrics/functional/multimodal/clip_score.py:79]: Yes, we should rather check for the version as well 🤔

File: docs/source/index.rst [hunk 12]
```diff
@@ -182,6 +182,14 @@ Or directly from conda
...
    nominal/*
 
+.. toctree::
+   :maxdepth: 2
+   :name: multimodal
+   :caption: Multimodal
+   :glob:
+
+   multimodal/*
+
```
Comment (stancld) [inline docs/source/index.rst:192]: Should be put in front of nominal.

**Review by stancld — APPROVED** [review]
Superb work! 💪

File: requirements/text_test.txt [hunk 13]
```diff
@@ -1,6 +1,6 @@
 jiwer>=2.3.0
 rouge-score>=0.0.4
 bert_score==0.3.10
-transformers>=4.4.0
+transformers>=4.10.0
```
Comment (Borda) [inline requirements/text_test.txt:4]:
```suggestion
transformers>4.4.0
```

File: src/torchmetrics/multimodal/clip_score.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+import torch
+from torch import Tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.multimodal.clip_score import _clip_score_update, _get_model_and_processor
+from torchmetrics.utilities.imports import _TRANSFORMERS_AVAILABLE, _TRANSFORMERS_GREATER_EQUAL_4_10
+
+if not (_TRANSFORMERS_AVAILABLE and _TRANSFORMERS_GREATER_EQUAL_4_10):
+    __doctest_skip__ = ["CLIPScore"]
```
Comment (Borda) [inline src/torchmetrics/multimodal/clip_score.py:24]:
```suggestion
from torchmetrics.utilities.imports import _TRANSFORMERS_AVAILABLE, _TRANSFORMERS_GREATER_EQUAL_4_10

if not _TRANSFORMERS_GREATER_EQUAL_4_10:
    __doctest_skip__ = ["CLIPScore"]
```

**Comment by Borda** [comment]
@SkafteNicki sorry, it was my bad to ask for var with min Trnformer version while they are now used only in one domain so there is no case of misconfiguration...

**Pull request merged**